forum/qa-plugin/user-activity-log/user-activity-log-search.php

```diff
@@ -0,0 +1,177 @@
+        if(!isset($_POST['condition']) || !isset($_POST['resultsCount'])){
+            header('Location: ../../');
+            exit;
+        }
+
+        if(qa_get_logged_in_level() >= QA_USER_LEVEL_EDITOR){
+            if(isset($this->searchArr)){
+                $this->dbSearch();
+                if(isset($this->dbResult)){
+                    return $request === 'user-activity-log-search';
+                }
+            }
+        }else{
+            return false;
+        }
+    }
+
+    public function process_request($request)
+    {
+        $qa_content=qa_content_prepare();
+        $qa_content['title']= qa_lang_html('user-activity-log/searchResults').$this->searchArr['request'];
+        
+        $qa_content['custom'] = $this->generateResultsTable();
+        $qa_content['navigation']['sub']=qa_admin_sub_navigation();
+        return $qa_content;
+    }
+
+    private function dbSearch()
+    {
+        $baseQuery = 'SELECT `datetime`, `ipaddress`, `handle`, `event` FROM qa_eventlog ';
+
+        if($this->searchArr['condition'] === 'username'){
+            $finalQuery = $baseQuery.'WHERE  handle = $';
+        }else if($this->searchArr['condition'] === 'type'){
+            $finalQuery = $baseQuery.'WHERE event = $';
+        }else if($this->searchArr['condition'] === 'ip'){
+            if(qa_get_logged_in_level() > QA_USER_LEVEL_EDITOR){
+                $finalQuery = $baseQuery.'WHERE ipaddress = $';
+            }   
+        }
+
+        if(!empty($this->searchArr['date'])){
+            $dateArr = explode('-', $this->searchArr['date']);
+
+            if(isset($dateArr[0]) && is_numeric($dateArr[0])){
+                $date = $dateArr[0];
+                if(isset($dateArr[1]) && is_numeric($dateArr[1])){
+                    $date = $date.'-'.$dateArr[1];
+                    if(isset($dateArr[2])){
+                        if(strlen($dateArr[2]) > 2){
+                            [$day, $hours] = explode(' ', $dateArr[2]);
+                            $dateArr[2] = $day;
+                            $dateArr[3] = $hours;
+                        }
+                        
+                        $date = $date.'-'.$dateArr[2];
+                    }
+                }
+                $finalQuery = $finalQuery.'AND `datetime` LIKE $';
+            }
+
+        }
+        
+        
+        if(!is_numeric($this->searchArr['resultsCount'])){
+            $this->dbResult = [];
+            return 0;
+        }
+        
+        $finalQuery = $finalQuery.' ORDER BY `datetime` DESC LIMIT #';
+        if(strpos($finalQuery, 'LIKE')){
+            $result = qa_db_query_sub($finalQuery, $this->searchArr['request'], $this->searchArr['date']."%", $this->searchArr['resultsCount']);
+        }else{
+            $result = qa_db_query_sub($finalQuery, $this->searchArr['request'], $this->searchArr['resultsCount']);
+        }
+       
+        if(mysqli_num_rows($result) <= $this->searchArr['resultsCount']){
```

Comment:
Chyba Q2A ma do tego funkcje, np. `qa_db_read_all_assoc()` może się nada, aby nie trzeba było używać natywnych z PHP. Oczywiście w tych też nic złego nie ma, ale jak już Q2A leci w miarę jednym to można by i tu starać się ich użyć.